tests/test_basic.c

```diff
@@ -0,0 +1,237 @@
+/* vim:set ft=c ts=2 sw=2 sts=2 et cindent: */
+/*
+ * Copyright 2014 Michael Steinert
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "amqp.h"
+#include "amqp_tcp_socket.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef _WIN32
+#include <WinSock2.h>
+#else
+#include <sys/time.h>
+#endif
+
+#define assert(x)                                    \
+  {                                                  \
+    if (!(x)) {                                      \
+      fprintf(stderr, "Assertion failed: %s\n", #x); \
+      abort();                                       \
+    }                                                \
+  }
+
+static const int fixed_channel_id = 1;
+static const char test_queue_name[] = "test_queue";
+
+amqp_connection_state_t setup_connection_and_channel(void) {
+  amqp_connection_state_t connection_state_ = amqp_new_connection();
+
+  amqp_socket_t *socket = amqp_tcp_socket_new(connection_state_);
+  assert(socket);
+
+  int rc = amqp_socket_open(socket, "localhost", AMQP_PROTOCOL_PORT);
+  assert(rc == AMQP_STATUS_OK);
+
+  amqp_rpc_reply_t rpc_reply = amqp_login_with_properties(
+      connection_state_, "/", 1, AMQP_DEFAULT_FRAME_SIZE, 0, NULL,
+      AMQP_SASL_METHOD_PLAIN, "guest", "guest");
+  assert(rpc_reply.reply_type == AMQP_RESPONSE_NORMAL);
+
+  amqp_channel_open_ok_t *res =
+      amqp_channel_open(connection_state_, fixed_channel_id);
+  assert(res != NULL);
+
+  return connection_state_;
+}
+
+void close_channel_and_connection(amqp_connection_state_t connection_state_) {
+  amqp_rpc_reply_t rpc_reply =
+      amqp_channel_close(connection_state_, 1, AMQP_REPLY_SUCCESS);
```

Comment:
If you're tearing down the whole connection, don't bother tearing down the channel.

If you want code coverage to explicitly cover amqp_channel_close, write a test that covers that.

Reply:
Hm, I agree there should be tests without and with closing the channel. However, don't you think the graceful way includes closing the channel first? Also, not closing the channel might mask channel exceptions.

Reply:
No need to be graceful about closing channels.  AMQP/RabbitMQ broker does not expect clients to close channels before shutting down the connection. Its a bit of a waste of network time if the broker will clean things up anyway.

Depending on what fails, you may or may not catch it here (the broker gives some leeway about closing a channel that has already been closed, in the event that both the client and the broker decide to close the channel at the same time).

If you want to use the channel-close as a mechanism for catching errors - it belongs in the method under test, not in a cleanup function.

Reply:
Ok